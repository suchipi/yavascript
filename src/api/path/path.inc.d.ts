/** An object that represents a filesystem path. */
declare class Path {
  /** The character used to separate path segments on this OS. */
  static readonly OS_SEGMENT_SEPARATOR: "/" | "\\";

  /**
   * The character used to separate entries within the PATH environment
   * variable on this OS.
   */
  static readonly OS_ENV_VAR_SEPARATOR: ":" | ";";

  /**
   * A list of suffixes that could appear in the filename for a program on the
   * current OS. For instance, on Windows, programs often end with ".exe".
   *
   * On Unix-like OSes, this is empty, On Windows, it's based on `env.PATHEXT`.
   */
  static readonly OS_PROGRAM_EXTENSIONS: ReadonlySet<string>;

  /** Split one or more path strings into an array of path segments. */
  static splitToSegments(inputParts: Array<string> | string): Array<string>;

  /**
   * Search the provided path string or strings for a path separator character,
   * and return it. If none is found, return `fallback`, which defaults to the
   * OS's path segment separator.
   */
  static detectSeparator<Fallback extends string | null = string>(
    input: Array<string> | string,
    // @ts-ignore might be instantiated with a different subtype
    fallback: Fallback = Path.OS_SEGMENT_SEPARATOR
  ): string | Fallback;

  /** Join together one or more paths. */
  static join(...inputs: Array<string | Path | Array<string | Path>>): Path;

  /**
   * Turns the input path(s) into an absolute path by resolving all `.` and `..`
   * segments, using `pwd()` as a base dir to use when resolving leading `.` or
   * `..` segments.
   */
  static resolve(...inputs: Array<string | Path | Array<string | Path>>): Path;

  /**
   * Concatenates the input path(s) and then resolves all non-leading `.` and
   * `..` segments.
   */
  static normalize(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): Path;

  /**
   * Return whether the provided path is absolute; that is, whether it
   * starts with either `/` or a drive letter (ie `C:`).
   */
  static isAbsolute(path: string | Path): boolean;

  /**
   * An array of the path segments that make up this path.
   *
   * For `/tmp/foo.txt`, it'd be `["", "tmp", "foo.txt"]`.
   *
   * For `C:\something\somewhere.txt`, it'd be `["C:", "something", "somewhere.txt"]`.
   */
  segments: Array<string>;

  /**
   * The path separator that should be used to turn this path into a string.
   *
   * Will be either `/` or `\`.
   */
  separator: string;

  /** Create a new Path object using the provided input(s). */
  constructor(...inputs: Array<string | Path | Array<string | Path>>);

  /** Create a new Path object using the provided segments and separator. */
  static from(segments: Array<string>, separator: string): Path;

  /**
   * Create an absolute path by `concat`ting `subpaths` onto this Path (which is
   * presumed to be an absolute path) and then using `normalize()` on the
   * result. If the result is not an absolute path, an error will be thrown.
   */
  resolve(...subpaths: Array<string | Path>): Path;

  /**
   * Resolve all non-leading `.` and `..` segments in this path.
   */
  normalize(): Path;

  /**
   * Create a new Path by appending additional path segments onto the end of
   * this Path's segments.
   *
   * The returned path will use this path's separator.
   */
  concat(...other: Array<string | Path | Array<string | Path>>): Path;

  /**
   * Return whether this path is absolute; that is, whether it starts with
   * either `/`, `\`, or a drive letter (ie `C:`).
   */
  isAbsolute(): boolean;

  /**
   * Make a second Path object containing the same segments and separator as
   * this one.
   */
  clone(): this;

  /**
   * Express this path relative to `dir`.
   *
   * @param dir - The directory to create a new path relative to.
   * @param options - Options that affect the resulting path.
   */
  relativeTo(
    dir: Path | string,
    options?: {
      /**
       * Defaults to false. When true, a leading `./` will be omitted from the
       * path, if present. Note that a leading `../` will never be omitted.
       */
      noLeadingDot?: boolean;
    }
  ): Path;

  /**
   * Turn this path into a string by joining its segments using its separator.
   */
  toString(): string;

  /**
   * Alias for `toString`; causes Path objects to be serialized as strings when
   * they (or an object referencing them) are passed into JSON.stringify.
   */
  toJSON(): string;

  /**
   * Return the final path segment of this path. If this path has no path
   * segments, the empty string is returned.
   */
  basename(): string;

  /**
   * Return the trailing extension of this path. The `options` parameter works
   * the same as the global `extname`'s `options` parameter.
   */
  extname(options?: { full?: boolean }): string;
}
