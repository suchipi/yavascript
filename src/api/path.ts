import * as os from "os";
import { makeErrorWithProperties } from "../error-with-properties";

function validateSegments(segments: Array<string>): Array<string> {
  return segments.filter((part, index) => {
    // first part can be "" to represent left side of root "/"
    if (index === 0) return true;
    return Boolean(part);
  });
}

export class Path {
  static OS_SEGMENT_SEPARATOR = os.platform === "win32" ? "\\" : "/";
  static OS_ENV_VAR_SEPARATOR = os.platform === "win32" ? ";" : ":";

  static splitToSegments(inputParts: Array<string> | string): Array<string> {
    if (!Array.isArray(inputParts)) {
      inputParts = [inputParts];
    }

    return validateSegments(
      inputParts.map((part) => part.split(/(?:\/|\\)+/g)).flat(1)
    );
  }

  static detectSeparator<Fallback extends string | null = string>(
    input: Array<string> | string,
    // @ts-ignore might be instantiated with a different subtype
    fallback: Fallback = Path.OS_SEGMENT_SEPARATOR
  ): string | Fallback {
    let testStr = input;
    if (Array.isArray(input)) {
      testStr = input.join("|");
    }

    for (const char of testStr) {
      if (char === "/") {
        return "/";
      } else if (char === "\\") {
        return "\\";
      }
    }

    return fallback;
  }

  static join(...inputs: Array<string | Path | Array<string | Path>>): string {
    return new Path(...inputs).toString();
  }

  static resolve(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): string {
    return new Path(...inputs).resolve().toString();
  }

  static normalize(
    ...inputs: Array<string | Path | Array<string | Path>>
  ): string {
    return new Path(...inputs).normalize().toString();
  }

  static isAbsolute(path: string): boolean {
    return new Path(path).isAbsolute();
  }

  static tag(
    _strings: TemplateStringsArray,
    ..._values: ReadonlyArray<string | Path | Array<string | Path>>
  ) {
    const strings = [..._strings];
    const values = [..._values];

    const zipped: Array<string | Path | Array<string | Path>> = [];

    let sourceIsStrings = true;
    while (strings.length + values.length > 0) {
      const next = sourceIsStrings ? strings.shift() : values.shift();
      if (next != null) {
        zipped.push(next);
      }
      sourceIsStrings = !sourceIsStrings;
    }

    if (zipped[0] === "") {
      zipped.shift();
    }

    return new Path(...zipped);
  }

  static tagUsingBase(dir: string | Path): typeof Path.tag {
    const ret = (strings, ...values) => {
      const stringsClone = Object.assign([...strings], {
        raw: [...strings.raw],
      });
      stringsClone[0] = Path.join(dir, strings[0]);
      stringsClone.raw[0] = Path.join(dir, strings.raw[0]);
      return Path.tag(stringsClone, ...values);
    };
    Object.defineProperty(ret, "name", {
      value: `tagUsingBase(${dir})`,
      configurable: true,
    });
    return ret;
  }

  segments: Array<string>;
  separator: string;

  constructor(...inputs: Array<string | Path | Array<string | Path>>) {
    const parts = inputs
      .flat(1)
      .map((part) => (typeof part === "string" ? part : part.segments))
      .flat(1);

    this.segments = Path.splitToSegments(parts);
    this.separator = Path.detectSeparator(parts);
  }

  static from(
    segments: Array<string>,
    separator: string = Path.OS_SEGMENT_SEPARATOR
  ) {
    const path = new Path();
    path.segments = validateSegments(segments);
    path.separator = separator;
    return path;
  }

  resolve(from: string | Path = os.getcwd()): Path {
    if (this.isAbsolute()) {
      return this.normalize();
    }

    const fromPath = typeof from === "string" ? new Path(from) : from;
    const result = fromPath.concat(this).normalize();
    if (!result.isAbsolute()) {
      throw makeErrorWithProperties(
        `Could not resolve ${this.toString()} from ${fromPath.toString()}`,
        {
          this: this.toString(),
          from: fromPath.toString(),
        }
      );
    }
    return result;
  }

  normalize(): Path {
    // we clone this cause we're gonna mutate it
    const segments = [...this.segments];

    const newSegments: Array<string> = [];
    let currentSegment: string | undefined;
    while (segments.length > 0) {
      currentSegment = segments.shift();

      switch (currentSegment) {
        case ".": {
          if (newSegments.length === 0) {
            newSegments.push(currentSegment);
          }
          break;
        }
        case "..": {
          if (newSegments.length === 0) {
            newSegments.push(currentSegment);
          } else {
            newSegments.pop();
          }

          break;
        }
        default: {
          if (currentSegment != null) {
            newSegments.push(currentSegment);
          }
          break;
        }
      }
    }

    return Path.from(newSegments, this.separator);
  }

  concat(other: string | Path | Array<string | Path>): Path {
    const otherSegments = new Path(other).segments;
    return Path.from(this.segments.concat(otherSegments), this.separator);
  }

  isAbsolute(): boolean {
    const firstPart = this.segments[0];

    // empty first component indicates that path starts with leading slash
    if (firstPart === "") return true;

    // windows drive
    if (/^[A-Za-z]:/.test(firstPart)) return true;

    // TODO: windows UNC paths (not supported very well by these path APIs at all, yet)

    return false;
  }

  toString() {
    const result = this.segments.join(this.separator);
    if (result == "") {
      return "/";
    } else {
      return result;
    }
  }
}
