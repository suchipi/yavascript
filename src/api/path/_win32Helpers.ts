import * as os from "quickjs:os";

const WIN32_DRIVE_LETTER_REGEXP = /^[A-Za-z]:$/;

export function appendSlashIfWindowsDriveLetter(pathString: string): string {
  if (os.platform === "win32" && WIN32_DRIVE_LETTER_REGEXP.test(pathString)) {
    return pathString + Path.OS_SEGMENT_SEPARATOR;
  } else {
    return pathString;
  }
}
