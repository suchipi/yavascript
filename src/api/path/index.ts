import { Path } from "./path";
import "./help-text/apply-help-text";
export { Path };
