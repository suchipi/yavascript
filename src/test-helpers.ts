import { spawn, Options as SpawnOptions } from "first-base";
import child_process from "child_process";
import path from "path";
import print from "@suchipi/print";
import * as inspectOptions from "./inspect-options";

export const TMP = child_process
  .execSync("realpath /tmp", { encoding: "utf-8" })
  .trim();

export const rootDir = (...parts: Array<string>) =>
  path.resolve(__dirname, "..", ...parts);

function getBinaryPath(platform: string, arch: string) {
  if (platform === "win32") {
    return path.resolve(__dirname, "..", "bin", "windows", "yavascript.exe");
  } else if (platform === "darwin") {
    if (arch.startsWith("arm")) {
      return path.resolve(__dirname, "..", "bin", "darwin-arm", "yavascript");
    } else {
      return path.resolve(__dirname, "..", "bin", "darwin", "yavascript");
    }
  } else if (platform === "linux") {
    return path.resolve(__dirname, "..", "bin", "linux", "yavascript");
  } else {
    throw new Error("Unsupported platform: " + platform);
  }
}

var binaryPath = getBinaryPath(
  process.platform,
  process.env.YS_TESTS_ARCH || process.arch
);

export { binaryPath, getBinaryPath };

export type EvaluateResult = {
  stdout: string;
  stderr: string;
  code: number | null;
  error: boolean;
};

export async function evaluate(
  code: string,
  options: SpawnOptions = {}
): Promise<EvaluateResult> {
  const runContext = spawn(binaryPath, ["-e", code], options);
  await runContext.completion;
  return runContext.result;
}

export function inspect(value: any): string {
  // options to inspect here match what is given to console.log
  return print(value, inspectOptions.forPrint);
}

export function cleanOutput(input: string): string {
  return input
    .replace(/yavascript:\d+/g, "yavascript")
    .replace(/\/bin\/[^/]+\/yavascript/g, "/bin/.../yavascript")
    .replace(new RegExp(rootDir(), "g"), "<rootDir>");
}

export function cleanResult(input: EvaluateResult): EvaluateResult {
  return {
    ...input,
    stdout: cleanOutput(input.stdout),
    stderr: cleanOutput(input.stderr),
  };
}
