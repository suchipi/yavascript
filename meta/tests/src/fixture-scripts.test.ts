import fs from "fs";
import path from "path";
import { spawn } from "first-base";
import { binaryPath, cleanResult, rootDir } from "./test-helpers";

const scriptsDir = rootDir("meta/tests/fixtures/scripts");

const scripts = fs.readdirSync(scriptsDir);
for (const script of scripts) {
  test(script, async () => {
    const run = spawn(binaryPath, [script], { cwd: scriptsDir });
    await run.completion;
    expect(cleanResult(run.result)).toMatchSnapshot();
  });
}
