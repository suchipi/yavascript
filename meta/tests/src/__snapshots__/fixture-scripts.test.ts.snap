// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__filename-and-__dirname.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  __filename: "<rootDir>/meta/tests/fixtures/scripts/__filename-and-__dirname.js"
  __dirname: "<rootDir>/meta/tests/fixtures/scripts"
}
{
  __filename: "<rootDir>/meta/tests/fixtures/scripts/__filename-and-__dirname-2.js"
  __dirname: "<rootDir>/meta/tests/fixtures/scripts"
}
",
}
`;

exports[`__filename-and-__dirname-2.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`http-import-error.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: Failed to load module (httpStatusCode = 404, url = "http://suchipi.com/there-is-nothing-at-this-path")
  at somewhere
{
  httpStatusCode: 404
  url: "http://suchipi.com/there-is-nothing-at-this-path"
  httpResponseHeaders: <redacted>
  httpResponseBody: <redacted>
}
",
  "stdout": "",
}
`;

exports[`https-import-error.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: Failed to load module (httpStatusCode = 404, url = "https://suchipi.com/there-is-nothing-at-this-path")
  at somewhere
{
  httpStatusCode: 404
  url: "https://suchipi.com/there-is-nothing-at-this-path"
  httpResponseHeaders: <redacted>
  httpResponseBody: <redacted>
}
",
  "stdout": "",
}
`;

exports[`importing-from-npm-protocol.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  Null prototype
  Sealed
  
  $ApiFunctions: {
    get asType: Function "asType" {}
    get assertType: Function "assertType" {}
    get isOfType: Function "isOfType" {}
    get stringifyValue: Function "stringifyValue" {}
  }
  $BasicTypes: {
    get Array: Function "Array" {}
    get ArrayBuffer: Function "ArrayBuffer" {}
    get BigInt: Function "BigInt" {}
    get Boolean: Function "Boolean" {}
    get DataView: Function "DataView" {}
    get Date: Function "Date" {}
    get Error: Function "Error" {}
    get Float32Array: Function "Float32Array" {}
    get Float64Array: Function "Float64Array" {}
    get Function: Function "Function" {}
    get Infinity: Function "Infinity" {}
    get Int16Array: Function "Int16Array" {}
    get Int32Array: Function "Int32Array" {}
    get Int8Array: Function "Int8Array" {}
    get Map: Function "Map" {}
    get NaN: Function "NaN" {}
    get NegativeInfinity: Function "NegativeInfinity" {}
    get Number: Function "Number" {}
    get Object: Function "Object" {}
    get RegExp: Function "RegExp" {}
    get Set: Function "Set" {}
    get SharedArrayBuffer: Function "SharedArrayBuffer" {}
    get String: Function "String" {}
    get Symbol: Function "Symbol" {}
    get TypedArray: Function "TypedArray" {}
    get Uint16Array: Function "Uint16Array" {}
    get Uint32Array: Function "Uint32Array" {}
    get Uint8Array: Function "Uint8Array" {}
    get Uint8ClampedArray: Function "Uint8ClampedArray" {}
    get any: Function "any" {}
    get anyArray: Function "anyArray" {}
    get anyFunction: Function "anyFunction" {}
    get anyMap: Function "anyMap" {}
    get anyObject: Function "anyObject" {}
    get anySet: Function "anySet" {}
    get anyTypeValidator: Function "anyTypeValidator" {}
    get array: Function "array" {}
    get arrayOfAny: Function "arrayOfAny" {}
    get arrayOfUnknown: Function "arrayOfUnknown" {}
    get bigint: Function "bigint" {}
    get boolean: Function "boolean" {}
    get false: Function "false" {}
    get falsy: Function "falsy" {}
    get integer: Function "integer" {}
    get map: Function "map" {}
    get never: Function "never" {}
    get nonNullOrUndefined: Function "nonNullOrUndefined" {}
    get null: Function "null" {}
    get nullish: Function "nullish" {}
    get number: Function "number" {}
    get numberIncludingNanAndInfinities: Function "numberIncludingNanAndInfinities" {}
    get object: Function "object" {}
    get objectOrNull: Function "objectOrNull" {}
    get set: Function "set" {}
    get string: Function "string" {}
    get symbol: Function "symbol" {}
    get true: Function "true" {}
    get truthy: Function "truthy" {}
    get undefined: Function "undefined" {}
    get unknown: Function "unknown" {}
    get unknownFunction: Function "unknownFunction" {}
    get unknownMap: Function "unknownMap" {}
    get unknownObject: Function "unknownObject" {}
    get unknownSet: Function "unknownSet" {}
    get unknownTypeValidator: Function "unknownTypeValidator" {}
    get void: Function "void" {}
  }
  $TypeConstructors: {
    get and: Function "and" {}
    get arrayOf: Function "arrayOf" {}
    get exactBigInt: Function "exactBigInt" {}
    get exactNumber: Function "exactNumber" {}
    get exactString: Function "exactString" {}
    get exactSymbol: Function "exactSymbol" {}
    get hasClassName: Function "hasClassName" {}
    get hasToStringTag: Function "hasToStringTag" {}
    get instanceOf: Function "instanceOf" {}
    get intersection: Function "intersection" {}
    get mapOf: Function "mapOf" {}
    get mappingObjectOf: Function "mappingObjectOf" {}
    get maybe: Function "maybe" {}
    get objectStr: Function "objectStr" {}
    get objectWithOnlyTheseProperties: Function "objectWithOnlyTheseProperties" {}
    get objectWithProperties: Function "objectWithProperties" {}
    get optional: Function "optional" {}
    get or: Function "or" {}
    get partialObjectWithProperties: Function "partialObjectWithProperties" {}
    get record: Function "record" {}
    get setOf: Function "setOf" {}
    get stringMatching: Function "stringMatching" {}
    get symbolFor: Function "symbolFor" {}
    get tuple: Function "tuple" {}
    get union: Function "union" {}
  }
  Array: Function "array" {}
  ArrayBuffer: Function "ArrayBuffer" {}
  BigInt: Function "bigint" {}
  Boolean: Function "boolean" {}
  DataView: Function "DataView" {}
  Date: Function "Date" {}
  Error: Function "Error" {}
  Float32Array: Function "Float32Array" {}
  Float64Array: Function "Float64Array" {}
  Function: Function "unknownFunction" {}
  Int16Array: Function "Int16Array" {}
  Int32Array: Function "Int32Array" {}
  Int8Array: Function "Int8Array" {}
  Map: Function "map" {}
  NegativeInfinity: Function "NegativeInfinity" {}
  Number: Function "number" {}
  Object: Function "object" {}
  RegExp: Function "RegExp" {}
  Set: Function "set" {}
  SharedArrayBuffer: Function "SharedArrayBuffer" {}
  String: Function "string" {}
  Symbol: Function "Symbol" {}
  TypedArray: Function "TypedArray" {}
  Uint16Array: Function "Uint16Array" {}
  Uint32Array: Function "Uint32Array" {}
  Uint8Array: Function "Uint8Array" {}
  Uint8ClampedArray: Function "Uint8ClampedArray" {}
  __moduleExports: {
    get $ApiFunctions: Function "$ApiFunctions" {}
    get $BasicTypes: Function "$BasicTypes" {}
    get $TypeConstructors: Function "$TypeConstructors" {}
    get asType: Function "get" {}
    get assertType: Function "get" {}
    get isOfType: Function "get" {}
    get stringifyValue: Function "get" {}
    get Array: Function "get" {}
    get ArrayBuffer: Function "get" {}
    get BigInt: Function "get" {}
    get Boolean: Function "get" {}
    get DataView: Function "get" {}
    get Date: Function "get" {}
    get Error: Function "get" {}
    get Float32Array: Function "get" {}
    get Float64Array: Function "get" {}
    get Function: Function "get" {}
    get Infinity: Function "get" {}
    get Int16Array: Function "get" {}
    get Int32Array: Function "get" {}
    get Int8Array: Function "get" {}
    get Map: Function "get" {}
    get NaN: Function "get" {}
    get NegativeInfinity: Function "get" {}
    get Number: Function "get" {}
    get Object: Function "get" {}
    get RegExp: Function "get" {}
    get Set: Function "get" {}
    get SharedArrayBuffer: Function "get" {}
    get String: Function "get" {}
    get Symbol: Function "get" {}
    get TypedArray: Function "get" {}
    get Uint16Array: Function "get" {}
    get Uint32Array: Function "get" {}
    get Uint8Array: Function "get" {}
    get Uint8ClampedArray: Function "get" {}
    get any: Function "get" {}
    get anyArray: Function "get" {}
    get anyFunction: Function "get" {}
    get anyMap: Function "get" {}
    get anyObject: Function "get" {}
    get anySet: Function "get" {}
    get anyTypeValidator: Function "get" {}
    get array: Function "get" {}
    get arrayOfAny: Function "get" {}
    get arrayOfUnknown: Function "get" {}
    get bigint: Function "get" {}
    get boolean: Function "get" {}
    get false: Function "get" {}
    get falsy: Function "get" {}
    get integer: Function "get" {}
    get map: Function "get" {}
    get never: Function "get" {}
    get nonNullOrUndefined: Function "get" {}
    get null: Function "get" {}
    get nullish: Function "get" {}
    get number: Function "get" {}
    get numberIncludingNanAndInfinities: Function "get" {}
    get object: Function "get" {}
    get objectOrNull: Function "get" {}
    get set: Function "get" {}
    get string: Function "get" {}
    get symbol: Function "get" {}
    get true: Function "get" {}
    get truthy: Function "get" {}
    get undefined: Function "get" {}
    get unknown: Function "get" {}
    get unknownFunction: Function "get" {}
    get unknownMap: Function "get" {}
    get unknownObject: Function "get" {}
    get unknownSet: Function "get" {}
    get unknownTypeValidator: Function "get" {}
    get void: Function "get" {}
    get and: Function "get" {}
    get arrayOf: Function "get" {}
    get exactBigInt: Function "get" {}
    get exactNumber: Function "get" {}
    get exactString: Function "get" {}
    get exactSymbol: Function "get" {}
    get hasClassName: Function "get" {}
    get hasToStringTag: Function "get" {}
    get instanceOf: Function "get" {}
    get intersection: Function "get" {}
    get mapOf: Function "get" {}
    get mappingObjectOf: Function "get" {}
    get maybe: Function "get" {}
    get objectStr: Function "get" {}
    get objectWithOnlyTheseProperties: Function "get" {}
    get objectWithProperties: Function "get" {}
    get optional: Function "get" {}
    get or: Function "get" {}
    get partialObjectWithProperties: Function "get" {}
    get record: Function "get" {}
    get setOf: Function "get" {}
    get stringMatching: Function "get" {}
    get symbolFor: Function "get" {}
    get tuple: Function "get" {}
    get union: Function "get" {}
  }
  and: Function "O" {}
  any: Function "unknown" {}
  anyArray: Function "anyArray" {}
  anyFunction: Function "anyFunction" {}
  anyMap: Function "anyMap" {}
  anyObject: Function "anyObject" {}
  anySet: Function "anySet" {}
  anyTypeValidator: Function "anyTypeValidator" {}
  array: -> {root}.Array
  arrayOf: Function "m" {}
  arrayOfAny: Function "arrayOfAny" {}
  arrayOfUnknown: Function "arrayOfUnknown" {}
  asType: Function "d" {}
  assertType: Function "assertType" {}
  bigint: -> {root}.BigInt
  default: -> {root}.__moduleExports
  exactBigInt: Function "b" {}
  exactNumber: Function "v" {}
  exactString: Function "_" {}
  exactSymbol: Function "g" {}
  falsy: Function "falsy" {}
  hasClassName: Function "j" {}
  hasToStringTag: Function "N" {}
  instanceOf: Function "I" {}
  integer: Function "integer" {}
  intersection: -> {root}.and
  isOfType: Function "p" {}
  map: -> {root}.Map
  mapOf: Function "x" {}
  mappingObjectOf: Function "V" {}
  maybe: Function "P" {}
  never: Function "never" {}
  nonNullOrUndefined: Function "nonNullOrUndefined" {}
  nullish: Function "nullish" {}
  number: -> {root}.Number
  numberIncludingNanAndInfinities: Function "numberIncludingNanAndInfinities" {}
  object: -> {root}.Object
  objectOrNull: Function "objectOrNull" {}
  objectStr: Function "d" {}
  objectWithOnlyTheseProperties: Function "U" {}
  objectWithProperties: Function "k" {}
  optional: Function "M" {}
  or: Function "A" {}
  partialObjectWithProperties: Function "B" {}
  record: -> {root}.mappingObjectOf
  set: -> {root}.Set
  setOf: Function "E" {}
  string: -> {root}.String
  stringMatching: Function "D" {}
  stringifyValue: Function "c" {}
  symbol: -> {root}.Symbol
  symbolFor: Function "W" {}
  truthy: Function "truthy" {}
  tuple: Function "C" {}
  union: -> {root}.or
  unknown: Function "f" {}
  unknownFunction: -> {root}.Function
  unknownMap: Function "unknownMap" {}
  unknownObject: Function "unknownObject" {}
  unknownSet: Function "unknownSet" {}
  unknownTypeValidator: Function "unknownTypeValidator" {}
}
",
}
`;

exports[`importing-json.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  hello: "there"
}
",
}
`;

exports[`npm-proto-import-error.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: Failed to load module (httpStatusCode = 404, url = "https://cdn.skypack.dev/@suchipi/this-package-does-not-exist")
  at somewhere
{
  httpStatusCode: 404
  url: "https://cdn.skypack.dev/@suchipi/this-package-does-not-exist"
  httpResponseHeaders: <redacted>
  httpResponseBody: <redacted>
}
",
  "stdout": "",
}
`;

exports[`requiring-with-npm-proto.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  get $ApiFunctions: Function "$ApiFunctions" {}
  get $BasicTypes: Function "$BasicTypes" {}
  get $TypeConstructors: Function "$TypeConstructors" {}
  get asType: Function "get" {}
  get assertType: Function "get" {}
  get isOfType: Function "get" {}
  get stringifyValue: Function "get" {}
  get Array: Function "get" {}
  get ArrayBuffer: Function "get" {}
  get BigInt: Function "get" {}
  get Boolean: Function "get" {}
  get DataView: Function "get" {}
  get Date: Function "get" {}
  get Error: Function "get" {}
  get Float32Array: Function "get" {}
  get Float64Array: Function "get" {}
  get Function: Function "get" {}
  get Infinity: Function "get" {}
  get Int16Array: Function "get" {}
  get Int32Array: Function "get" {}
  get Int8Array: Function "get" {}
  get Map: Function "get" {}
  get NaN: Function "get" {}
  get NegativeInfinity: Function "get" {}
  get Number: Function "get" {}
  get Object: Function "get" {}
  get RegExp: Function "get" {}
  get Set: Function "get" {}
  get SharedArrayBuffer: Function "get" {}
  get String: Function "get" {}
  get Symbol: Function "get" {}
  get TypedArray: Function "get" {}
  get Uint16Array: Function "get" {}
  get Uint32Array: Function "get" {}
  get Uint8Array: Function "get" {}
  get Uint8ClampedArray: Function "get" {}
  get any: Function "get" {}
  get anyArray: Function "get" {}
  get anyFunction: Function "get" {}
  get anyMap: Function "get" {}
  get anyObject: Function "get" {}
  get anySet: Function "get" {}
  get anyTypeValidator: Function "get" {}
  get array: Function "get" {}
  get arrayOfAny: Function "get" {}
  get arrayOfUnknown: Function "get" {}
  get bigint: Function "get" {}
  get boolean: Function "get" {}
  get false: Function "get" {}
  get falsy: Function "get" {}
  get integer: Function "get" {}
  get map: Function "get" {}
  get never: Function "get" {}
  get nonNullOrUndefined: Function "get" {}
  get null: Function "get" {}
  get nullish: Function "get" {}
  get number: Function "get" {}
  get numberIncludingNanAndInfinities: Function "get" {}
  get object: Function "get" {}
  get objectOrNull: Function "get" {}
  get set: Function "get" {}
  get string: Function "get" {}
  get symbol: Function "get" {}
  get true: Function "get" {}
  get truthy: Function "get" {}
  get undefined: Function "get" {}
  get unknown: Function "get" {}
  get unknownFunction: Function "get" {}
  get unknownMap: Function "get" {}
  get unknownObject: Function "get" {}
  get unknownSet: Function "get" {}
  get unknownTypeValidator: Function "get" {}
  get void: Function "get" {}
  get and: Function "get" {}
  get arrayOf: Function "get" {}
  get exactBigInt: Function "get" {}
  get exactNumber: Function "get" {}
  get exactString: Function "get" {}
  get exactSymbol: Function "get" {}
  get hasClassName: Function "get" {}
  get hasToStringTag: Function "get" {}
  get instanceOf: Function "get" {}
  get intersection: Function "get" {}
  get mapOf: Function "get" {}
  get mappingObjectOf: Function "get" {}
  get maybe: Function "get" {}
  get objectStr: Function "get" {}
  get objectWithOnlyTheseProperties: Function "get" {}
  get objectWithProperties: Function "get" {}
  get optional: Function "get" {}
  get or: Function "get" {}
  get partialObjectWithProperties: Function "get" {}
  get record: Function "get" {}
  get setOf: Function "get" {}
  get stringMatching: Function "get" {}
  get symbolFor: Function "get" {}
  get tuple: Function "get" {}
  get union: Function "get" {}
}
",
}
`;

exports[`some-civet.civet 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "object
object
false
{
  $$typeof: Symbol(JSX.Element)
  type: "div"
  props: {
    class: "hi"
    disabled: true
    children: [
      "hi"
    ]
  }
  key: null
}
",
}
`;

exports[`some-coffeescript.coffee 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "object
object
false
",
}
`;

exports[`some-json.json 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`some-jsx.jsx 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: "a"
  props: null
  key: null
}
true
false
true
false
false
",
}
`;

exports[`some-tsx.tsx 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: "a"
  props: null
  key: null
}
true
false
true
false
false
",
}
`;

exports[`some-typescript.ts 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "undefined
object
path: <rootDir>/meta/tests/fixtures/scripts
Path { <rootDir>/meta/tests/fixtures/scripts }
",
}
`;

exports[`toplevel-await.coffee 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "4 Function "open" {}
",
}
`;

exports[`toplevel-await.js 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "4 Function "open" {}
",
}
`;

exports[`toplevel-await.ts 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: nope!
  at somewhere
",
  "stdout": "",
}
`;

exports[`toplevel-await.tsx 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  $$typeof: Symbol(JSX.Element)
  type: "a"
  props: null
  key: null
}
Function "open" {}
",
}
`;

exports[`toplevel-await-and-export-together.js 1`] = `
{
  "code": 1,
  "error": false,
  "stderr": "Error: You cannot use toplevel await and export in the same file. Either remove the toplevel await (by wrapping stuff with an immediately-invoked async function) or remove the exports.
  at somewhere
{
  filename: "<rootDir>/meta/tests/fixtures/scripts/toplevel-await-and-export-together.js"
}
",
  "stdout": "",
}
`;

exports[`type-coercion-via-assert.ts 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "no error message from hi String
Expected value of type string, but received null
Expected value of type string, but received 24
no error message from hi hi
Expected value of type exactString("hi"), but received "yes"
Expected value of type number, but received "something"
no error message from 42 Number
",
}
`;

exports[`type-coercion-via-is.ts 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "true
false
false
true
false
false
true
",
}
`;
